polar-core/src/kb.rs

```diff
@@ -75,11 +81,247 @@ impl KnowledgeBase {
+                template_param.parameter.value(),
+                template_param.specializer.as_ref().map(Term::value),
+                rule_param.parameter.value(),
+                rule_param.specializer.as_ref().map(Term::value),
+            ) {
+                (
+                    Value::Variable(_),
+                    Some(Value::Pattern(template_spec)),
+                    Value::Variable(_),
+                    Some(Value::Pattern(rule_spec)),
+                ) => {
+                    // If template and rule both have pattern specializers, then rule
+                    // specializer must match template specializer
+                    match (template_spec, rule_spec) {
+                        (
+                            Pattern::Instance(template_instance),
+                            Pattern::Instance(rule_instance),
+                        ) => {
+                            // if tags match, all template fields must match those in rule fields, otherwise false
+                            if template_instance.tag == rule_instance.tag {
+                                self.param_fields_match(
+                                    &template_instance.fields,
+                                    &rule_instance.fields,
+                                )
+                            // If tags don't match, then rule specializer must be a subclass of template specializer
+                            } else if let Some(template_class) =
+                                self.constants.get(&template_instance.tag)
+                            {
+                                if let Some(rule_mro) = self.mro.get(&rule_instance.tag) {
+                                    rule_mro.contains(template_class)
+                                } else {
+                                    false
+                                }
+                            } else {
+                                false
+                            }
+                        }
+                        (
+                            Pattern::Dictionary(template_fields),
+                            Pattern::Dictionary(rule_fields),
+                        )
+                        | (
+                            Pattern::Dictionary(template_fields),
+                            Pattern::Instance(InstanceLiteral {
+                                tag: _,
+                                fields: rule_fields,
+                            }),
+                        ) => self.param_fields_match(template_fields, rule_fields),
+                        (
+                            Pattern::Instance(InstanceLiteral {
+                                tag,
+                                fields: template_fields,
+                            }),
+                            Pattern::Dictionary(rule_fields),
+                        ) if tag == &sym!("Dictionary") => {
+                            self.param_fields_match(template_fields, rule_fields)
```

Comment:
Dictionary specializer doesn't actually work, right?

Reply:
No because of a bug.